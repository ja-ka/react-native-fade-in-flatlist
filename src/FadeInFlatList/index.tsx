import React, { FC } from 'react';
import { ReactElement, useCallback, useEffect, useRef } from 'react';
import { Animated, Easing, FlatListProps, ListRenderItemInfo, FlatList } from 'react-native';

const FadeInFlatList = <ItemT,>({
  itemsToFadeIn = 10,
  initialDelay = 0,
  durationPerItem = 50,
  parallelItems = 1,
  renderItem: originalRenderItem,
  ItemSeparatorComponent,
  ...props
}: FlatListProps<ItemT> & {
  itemsToFadeIn?: number;
  initialDelay?: number;
  durationPerItem?: number;
  parallelItems?: number;
}): ReactElement => {
  const value = useRef(new Animated.Value(0));

  const FadeInComponent: FC<{ index: number }> = useCallback(
    ({ index, children }): ReactElement => {
      const moveBy = (1 - 1 / parallelItems) * index;

      return (
        <Animated.View
          style={{
            opacity: value.current.interpolate({
              inputRange: [
                index - 1 - moveBy,
                index - moveBy,
                index + 1 - moveBy,
                index + 2 - moveBy,
              ],
              outputRange: [0, 0, 1, 1],
              extrapolate: 'clamp',
            }),
          }}>
          {children}
        </Animated.View>
      );
    },
    [],
  );

  const Separator: FC<{ index: number }> = useCallback(
    ({ index }): ReactElement | null =>
      ItemSeparatorComponent && index !== undefined ? (
        <FadeInComponent index={index}>
          <ItemSeparatorComponent />
        </FadeInComponent>
      ) : ItemSeparatorComponent ? (
        <ItemSeparatorComponent />
      ) : null,
    [],
  );

  const Item: FC<{ info: ListRenderItemInfo<ItemT> }> = useCallback(({ info }): ReactElement => {
    useEffect(() => {
      info.separators.updateProps('leading', { index: info.index });
    }, []);

    return <FadeInComponent index={info.index}>{originalRenderItem!(info)}</FadeInComponent>;
  }, []);

  const renderItem = useCallback(
    (info: ListRenderItemInfo<ItemT>): React.ReactElement | null =>
      info.index < itemsToFadeIn ? <Item info={info} /> : originalRenderItem!(info),
    [originalRenderItem, itemsToFadeIn],
  );

  useEffect(() => {
    value.current.setValue(0);

    Animated.timing(value.current, {
      toValue: itemsToFadeIn + 1,
      useNativeDriver: true,
      delay: initialDelay,
      duration: itemsToFadeIn * durationPerItem,
      easing: Easing.linear,
    }).start();
  }, [initialDelay, durationPerItem, itemsToFadeIn]);

  return (
    <FlatList
      {...props}
      renderItem={renderItem}
      ItemSeparatorComponent={ItemSeparatorComponent ? Separator : null}
    />
  );
};

export default FadeInFlatList;
