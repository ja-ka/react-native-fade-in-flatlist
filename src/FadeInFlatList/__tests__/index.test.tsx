import React, { FC, ReactElement, useEffect, useRef } from 'react';
import { FlatList, Text } from 'react-native';
import { render, act } from '@testing-library/react-native';
import FadeInFlatList from '../index';

const DATA: {id: number}[] = [{ id: 42 }, { id: 43 }];
const renderItem = ({ item }: any) => <Text>{item.id}</Text>;
const keyExtractor = (({ id }: { id: number }) => id.toString()) as any;

it('should render the component', async () => {
  const { toJSON } = render(
    <FadeInFlatList
      data={DATA}
      renderItem={renderItem}
      itemsToFadeIn={1}
      initialDelay={100}
      durationPerItem={100}
      keyExtractor={keyExtractor}
    />,
  );

  await act(async () => {
    // this is ok
  });

  expect(toJSON()).toMatchSnapshot();
});

it('uses separator', async () => {
  const Separator: FC = (): null => null;

  const { toJSON } = render(
    <FadeInFlatList
      data={DATA}
      renderItem={renderItem}
      itemsToFadeIn={1}
      initialDelay={100}
      durationPerItem={100}
      keyExtractor={keyExtractor}
      ItemSeparatorComponent={Separator}
    />,
  );

  await act(async () => {
    // this is ok
  });

  expect(toJSON()).toMatchSnapshot();
});

it('should use defaults', async () => {
  const { toJSON } = render(
    <FadeInFlatList data={DATA} renderItem={renderItem} keyExtractor={keyExtractor} />,
  );

  await act(async () => {
    // this is ok
  });

  expect(toJSON()).toMatchSnapshot();
});

it('returns reference to ref', async () => {
  let called = 0;

  const MyComponent = (): ReactElement => {
    const listRef = useRef<FlatList>();

    useEffect(() => {
      called++;

      // @ts-expect-error possibly undefined
      listRef.current.scrollToEnd();
    }, []);

    return (
      <FadeInFlatList ref={listRef} data={DATA} renderItem={renderItem} keyExtractor={keyExtractor} />
    );
  };

  render(<MyComponent />);

  await act(async () => {
    // this is ok
  });

  expect(called).toBe(1);
});