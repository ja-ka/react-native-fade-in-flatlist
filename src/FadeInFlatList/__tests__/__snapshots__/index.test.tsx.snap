// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the component 1`] = `
<RCTScrollView
  ItemSeparatorComponent={null}
  data={
    Array [
      Object {
        "id": 42,
      },
      Object {
        "id": 43,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "opacity": 0,
          }
        }
      >
        <Text>
          42
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text>
        43
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`should use defaults 1`] = `
<RCTScrollView
  ItemSeparatorComponent={null}
  data={
    Array [
      Object {
        "id": 42,
      },
      Object {
        "id": 43,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "opacity": 0,
          }
        }
      >
        <Text>
          42
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "opacity": 0,
          }
        }
      >
        <Text>
          43
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`uses separator 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  data={
    Array [
      Object {
        "id": 42,
      },
      Object {
        "id": 43,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "opacity": 0,
          }
        }
      >
        <Text>
          42
        </Text>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Text>
        43
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
